corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,372 @@
+import hashlib
+from abc import abstractmethod, ABCMeta, abstractproperty
+from datetime import datetime
+
+import six
+from couchdbkit import ResourceNotFound
+
+from corehq.util.couch_helpers import PaginateViewLogHandler, paginate_view
+
+
+class ResumableDocsByTypeIterator(object):
+    """Perform one-time resumable iteration over documents by type
+
+    Iteration can be efficiently stopped and resumed. The iteration may
+    omit documents that are added after the iteration begins or resumes
+    and may include deleted documents.
+
+    Example usage
+    > iterator = ResumableDocsByTypeIterator(db, doc_types, 'my-key')
+    > for doc in iterator:
+        success = process_doc(doc)
+        if success:
+            iterator.mark_doc_processed(doc)
+        else:
+            iterator.retry(doc)
+    > print iterator.progress_info
+    > iterator.discard_state()
+
+    :param db: Couchdb database.
+    :param doc_types: A list of doc type names to iterate on.
+    :param iteration_key: A unique key identifying the iteration. This
+    key will be used in combination with `doc_types` to maintain state
+    about an iteration that is in progress. The state will be maintained
+    indefinitely unless it is removed with `discard_state()`.
+    :param chunk_size: Number of documents to yield before updating the
+    iteration checkpoint. In the worst case about this many documents
+    that were previously yielded may be yielded again if the iteration
+    is stopped and later resumed.
+    :param maintain_retry_state: True if the retry state of documents should be maintained
```

Comment:
Not understanding the need for this. This was previously possible by stopping (breaking) the iteration without completing it (like what would happen if `TooManyRetries` was raised and not handled inside the loop). I have an idea of a way to achieve this without adding an extra parameter or `mark_doc_processed` method. I'll add a comment to the code below...